### DISCUSSION TITLE
[Plugin] How to nest dark mode styles when registering classes

### DISCUSSION THREAD

**Discussion by mrcaidev** [tailwindlabs/tailwindcss#10460] [Help] [answered]
I want to register a component class in my plugin - let's call it `bg-custom` - which should equal to `bg-gray-100` in light mode, and `bg-gray-900` in dark mode.

I know this can easily be done with CSS:

```css
@layer components {
  .bg-custom {
    @apply bg-gray-100 dark:bg-gray-900
  }
}
```

But actually I was writing a plugin to port whole another color palette to Tailwind CSS, so I really need to find a way to do this with CSS-in-JS syntax.

Currently I've figured out to do this:

```js
const plugin = require("tailwindcss/plugin");

module.exports = {
  // ...
  darkMode: "class",
  plugins: [
    plugin(({ addComponents }) => {
      addComponents({
        ".bg-custom": {
          backgroundColor: "#f9fafb",
          ".dark &": {
            backgroundColor: "#111827",
          },
        },
      });
    }),
  ],
};
```

This works. But `.dark .bg-custom` will be bundled even if never used, see [reproduction](https://github.com/mrcaidev/tailwind-plugin-mre).

So I'm wondering what is the correct way to do this. Thank you!

**Comment by wongjn** [comment]
I believe that is a correct way of doing it. Indeed `.dark .bg-custom` may still be generated. You could convert your plugin to use [`plugin.withOptions()`](https://tailwindcss.com/docs/plugins#exposing-options) and have an option that can disable the `.dark` from added.

**Reply by mrcaidev** [reply]
Thanks for the suggestion! `withOptions` doc is helpful.

However, it does not make sense to disable `.dark`, because a large part of my original intention to write the plugin was to automatically support dark mode.

**Reply by mrcaidev** [reply]
And what if the user opens that option? The unused classes are still bundled.

It is really frustrating to see a 5KB CSS bloated to 20+KB 😭

**Comment by mrcaidev** [comment, marked as answer]
Perfect solution in #10468. Really helpful! ❤️
